File: webdisplay/backend/video_timestamp_mapper.py
"""
Video timestamp mapping utility
Converts flight data timestamps to YouTube video times using the timestamp mapping file
"""
import json
from pathlib import Path
from bisect import bisect_left

class VideoTimestampMapper:
    """Maps flight data timestamps to video times"""
    
    def __init__(self, timestamp_map_file=None):
        self.timestamps = []  # List of (data_timestamp, video_time) tuples
        self.data_timestamps = []  # Sorted list of data timestamps for binary search
        self.video_times = []  # Corresponding video times
        
        if timestamp_map_file and Path(timestamp_map_file).exists():
            self.load_mapping(timestamp_map_file)
    
    def load_mapping(self, timestamp_map_file):
        """Load timestamp mapping from JSON file"""
        try:
            with open(timestamp_map_file, 'r') as f:
                mapping = json.load(f)
                # Create sorted list of (data_timestamp, video_time) tuples
                self.timestamps = sorted([(entry['data_timestamp'], entry['video_time']) 
                                         for entry in mapping])
                self.data_timestamps = [ts[0] for ts in self.timestamps]
                self.video_times = [ts[1] for ts in self.timestamps]
                # Also create sorted list by video_time for reverse lookup
                self.timestamps_by_video = sorted(self.timestamps, key=lambda x: x[1])
                self.video_times_sorted = [ts[1] for ts in self.timestamps_by_video]
                self.data_timestamps_by_video = [ts[0] for ts in self.timestamps_by_video]
                return True
        except Exception as e:
            print(f"Error loading video timestamp mapping: {e}")
            return False
    
    def data_to_video_time(self, data_timestamp):
        """
        Convert flight data timestamp to video time using interpolation
        
        Args:
            data_timestamp: Flight data timestamp in seconds
            
        Returns:
            Video time in seconds, or None if mapping not available
        """
        if not self.timestamps:
            return None
        
        # Binary search for closest timestamp
        idx = bisect_left(self.data_timestamps, data_timestamp)
        
        # Handle edge cases
        if idx == 0:
            # Before first timestamp - use first mapping
            return self.video_times[0]
        elif idx >= len(self.timestamps):
            # After last timestamp - use last mapping
            return self.video_times[-1]
        
        # Interpolate between two points
        prev_data_ts, prev_video_ts = self.timestamps[idx - 1]
        next_data_ts, next_video_ts = self.timestamps[idx]
        
        # If exact match
        if prev_data_ts == data_timestamp:
            return prev_video_ts
        if next_data_ts == data_timestamp:
            return next_video_ts
        
        # Linear interpolation
        if next_data_ts == prev_data_ts:
            return prev_video_ts
        
        ratio = (data_timestamp - prev_data_ts) / (next_data_ts - prev_data_ts)
        video_time = prev_video_ts + ratio * (next_video_ts - prev_video_ts)
        
        return video_time
    
    def video_to_data_time(self, video_time):
        """
        Convert video time to flight data timestamp using interpolation (reverse mapping)
        
        Args:
            video_time: Video time in seconds
            
        Returns:
            Flight data timestamp in seconds, or None if mapping not available
        """
        if not self.timestamps or not hasattr(self, 'timestamps_by_video'):
            return None
        
        # Binary search for closest video time
        idx = bisect_left(self.video_times_sorted, video_time)
        
        # Handle edge cases
        if idx == 0:
            # Before first video time - use first mapping
            return self.data_timestamps_by_video[0]
        elif idx >= len(self.timestamps_by_video):
            # After last video time - use last mapping
            return self.data_timestamps_by_video[-1]
        
        # Interpolate between two points
        prev_data_ts, prev_video_ts = self.timestamps_by_video[idx - 1]
        next_data_ts, next_video_ts = self.timestamps_by_video[idx]
        
        # If exact match
        if prev_video_ts == video_time:
            return prev_data_ts
        if next_video_ts == video_time:
            return next_data_ts
        
        # Linear interpolation
        if next_video_ts == prev_video_ts:
            return prev_data_ts
        
        ratio = (video_time - prev_video_ts) / (next_video_ts - prev_video_ts)
        data_timestamp = prev_data_ts + ratio * (next_data_ts - prev_data_ts)
        
        return data_timestamp
    
    def is_available(self):
        """Check if mapping is available"""
        return len(self.timestamps) > 0

File: flight-display/server/videoTimestampMapper.js
/**
 * Video timestamp mapping utility
 * Converts flight data timestamps to YouTube video times using the timestamp mapping file
 */

const fs = require('fs');
const path = require('path');

class VideoTimestampMapper {
    /**
     * Initialize mapper with timestamp map file
     * @param {string} timestampMapFile - Path to JSON mapping file
     */
    constructor(timestampMapFile = null) {
        this.timestamps = []; // List of [data_timestamp, video_time] pairs
        this.dataTimestamps = []; // Sorted list of data timestamps for binary search
        this.videoTimes = []; // Corresponding video times
        this.timestampsByVideo = []; // Sorted by video time for reverse lookup
        this.videoTimesSorted = [];
        this.dataTimestampsByVideo = [];

        if (timestampMapFile && fs.existsSync(timestampMapFile)) {
            this.loadMapping(timestampMapFile);
        }
    }

    /**
     * Load timestamp mapping from JSON file
     */
    loadMapping(timestampMapFile) {
        try {
            const fileContent = fs.readFileSync(timestampMapFile, 'utf-8');
            const mapping = JSON.parse(fileContent);

            // Create sorted list of [data_timestamp, video_time] pairs
            this.timestamps = mapping
                .map(entry => [entry.data_timestamp, entry.video_time])
                .sort((a, b) => a[0] - b[0]);

            this.dataTimestamps = this.timestamps.map(ts => ts[0]);
            this.videoTimes = this.timestamps.map(ts => ts[1]);

            // Also create sorted list by video_time for reverse lookup
            this.timestampsByVideo = [...this.timestamps].sort((a, b) => a[1] - b[1]);
            this.videoTimesSorted = this.timestampsByVideo.map(ts => ts[1]);
            this.dataTimestampsByVideo = this.timestampsByVideo.map(ts => ts[0]);

            return true;
        } catch (error) {
            console.log(`Error loading video timestamp mapping: ${error}`);
            return false;
        }
    }

    /**
     * Binary search helper
     */
    _bisectLeft(arr, value) {
        let left = 0;
        let right = arr.length;
        while (left < right) {
            const mid = Math.floor((left + right) / 2);
            if (arr[mid] < value) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }

    /**
     * Convert flight data timestamp to video time using interpolation
     * @param {number} dataTimestamp - Flight data timestamp in seconds
     * @returns {number|null} Video time in seconds, or null if mapping not available
     */
    dataToVideoTime(dataTimestamp) {
        if (this.timestamps.length === 0) {
            return null;
        }

        // Binary search for closest timestamp
        let idx = this._bisectLeft(this.dataTimestamps, dataTimestamp);

        // Handle edge cases
        if (idx === 0) {
            return this.videoTimes[0];
        } else if (idx >= this.timestamps.length) {
            return this.videoTimes[this.videoTimes.length - 1];
        }

        // Interpolate between two points
        const [prevDataTs, prevVideoTs] = this.timestamps[idx - 1];
        const [nextDataTs, nextVideoTs] = this.timestamps[idx];

        // If exact match
        if (prevDataTs === dataTimestamp) {
            return prevVideoTs;
        }
        if (nextDataTs === dataTimestamp) {
            return nextVideoTs;
        }

        // Linear interpolation
        if (nextDataTs === prevDataTs) {
            return prevVideoTs;
        }

        const ratio = (dataTimestamp - prevDataTs) / (nextDataTs - prevDataTs);
        const videoTime = prevVideoTs + ratio * (nextVideoTs - prevVideoTs);

        return videoTime;
    }

    /**
     * Convert video time to flight data timestamp using interpolation (reverse mapping)
     * @param {number} videoTime - Video time in seconds
     * @returns {number|null} Flight data timestamp in seconds, or null if mapping not available
     */
    videoToDataTime(videoTime) {
        if (this.timestamps.length === 0 || this.timestampsByVideo.length === 0) {
            return null;
        }

        // Binary search for closest video time
        let idx = this._bisectLeft(this.videoTimesSorted, videoTime);

        // Handle edge cases
        if (idx === 0) {
            return this.dataTimestampsByVideo[0];
        } else if (idx >= this.timestampsByVideo.length) {
            return this.dataTimestampsByVideo[this.dataTimestampsByVideo.length - 1];
        }

        // Interpolate between two points
        const [prevDataTs, prevVideoTs] = this.timestampsByVideo[idx - 1];
        const [nextDataTs, nextVideoTs] = this.timestampsByVideo[idx];

        // If exact match
        if (prevVideoTs === videoTime) {
            return prevDataTs;
        }
        if (nextVideoTs === videoTime) {
            return nextDataTs;
        }

        // Linear interpolation
        if (nextVideoTs === prevVideoTs) {
            return prevDataTs;
        }

        const ratio = (videoTime - prevVideoTs) / (nextVideoTs - prevVideoTs);
        const dataTimestamp = prevDataTs + ratio * (nextDataTs - prevDataTs);

        return dataTimestamp;
    }

    /**
     * Check if mapping is available
     */
    isAvailable() {
        return this.timestamps.length > 0;
    }
}

module.exports = VideoTimestampMapper;

